### ISSUE TITLE
Changes in team drives get destroyed randomly

### ISSUE THREAD

**Issue by thomaskuntzz** [xwiki-labs/cryptpad#544]
**Describe the bug**
I have a team, with a team drive, with files in it, organized into folders. Sometimes (a few times a week), files that had been created recently disappear from the team drive and the ones that remain in the team drive (the ones that got created longer ago) see themselves renames to how they were named in the team drive a few days ago. All of this happens without anyone on the team having taken any cation for such a thing to occur. **It's as if the metadata of the team drive (location and name of the files _inside of the team drive structure_) was rolled back to its state one or two days before, but not the files themselves (the actual content of the files, their history, and their real name, the one you see when you open it).**

The files themselves do not get deleted, nor do their changes get reverted. As I have the URLs of the files that disappeared in my browser history, I can visit them, select "Add to my drive" and then from my drive add them back to the team drive where they were before, and finally put them back into place... Until the bug strikes again. (So this is a very annoying issue)

NB: When considering a file stored in a team drive, I see that there are two potentially different names: the name the name the file has _in the drive_, and the one it has "in itself" (the one you can see on the top left corner when you open the document)

Let's take a simple example. I have one file, "file1", created at t=0.  Let's say at t=1 I rename "file1" into "fileA" and make some changes to it. At t=2 I create a file named "file2". At t=3 I make changes to both of these files. And at t=4, when the bug occurs, the following occurs:

- "fileA" gets renamed into "file1" _inside the team drive_, but not in the document itself
- "file2" disappears from the team drive but can still be accessed from its URL
- both files' content remains the same as on t=3

It's **as if the metadata of the team drive (the file structure) got rolled back to t=0 but not the files itself**.

**Where did it happen?**
The issue occurred on CryptPad.fr

**To Reproduce**
If the conditions are there (some files in a team drive being modified and created over time), the bug happens for no apparent reason. I've been using CryptPad for my team for a few weeks or so and in the pas week it happened at least three times. So I would say it occurs every two or three days, but randomly and for no apparent reason.

**Expected behavior**
Well obviously nothing should change on its own...

**Screenshots**
No screenshots

**Browser (please complete the following information):**
 - OS: Ubuntu 20.04
 - Browser: Firefox
 - Version: it occurred with both versions 75 and 76
 - Extensions installed: UBlock Origin, PrivacyBadger, HTTPS Everywhere, Alternate Tube Redirector, Facebook Container
 - Browser tweaks: firefox "Enhanced Tracking Protection"
**Update:** Also happenned when I was on Windows 10 with Brave Browser (but with all the Brave blocking stuff disabled for CryptPad), so the issue is probable not due to an OS/Browser compatibilty problem

I tried to disable all my extensions, to the exception of Facebook Container, for CryptPad.fr ("disable for this site") and the issue still occurred.

NB: The bug might also be caused by someone else on the team (we're three people), but there's no real way to know...

PS: Thanks for all your amazing work with CryptPad! :heart:

PPS: Please tell me (if you can) if there might be a solution to be found in the coming days, otherwise I will have to move to another tool very quickly because I can't really get work done under those conditions

**Comment by ansuz** [comment]
I don't know how interested you are in the details of this problem, but I'll provide an overview in case anyone else experiences the problem and wants a better understanding of it. I'll follow that up with some steps that should help prevent it going forward.

User and team drives are implemented in a similar way to CryptPad's real-time editors. Whenever you make a change your browser generates a "patch" which is basically an instruction to transform the document from the previous state to what you have in your browser. When two or more users make changes at the same time there is a process which resolves the differences between the two patches so that everyone ends up with the same content. If the patches cannot be resolved then the document ends up with two parallel histories. Whichever branch of the history is the longest gets adopted as the "best branch". This whole process is done by the editors themselves, not the server, since all the patches are encrypted and the server doesn't have the relevant information to resolve concurrent patches.

The problem with this behaviour is that a document's history only ever gets longer. If we didn't do something to manage this unbounded growth then you would need to load every patch every time you loaded a drive or document. We avoid this with "checkpoints", which are a complete snapshot of the document at a specific point with a special format which indicates to the server that that patch is a checkpoint. When you try to load a drive or real-time document your browser asks the server for the most recent checkpoints and all the following patches. As long as everything is stored in the database in the right order then the server will provide a snapshot and all the amendments which bring a newly connected user up to date with all the other users.

We recently introduced some new server features to more tasks in parallel to handle in the increased load we've been receiving with more people working and studying from home. Unfortunately, this introduced a regression which made it possible for messages to be written to the database out of order. This in turn made it possible for users to miss a patch against a checkpoint, which has the ultimate result of an entire chain of modifications being considered invalid by clients without a more complete view of the history. In practice it then becomes possible for two sets of users to receive but ignore each other's messages and thus to witness different branches of history.

I believe what happened in your case is that two or more users have ended up with divergent versions of your team drive. As they perform actions like moving items in their drive or opening files and changing their access time they've caused one branch to suddenly become longer than another. The result is that the users that have a complete view of both branches abandon their present state and switch branches to the new "best branch", seen from your perspective as a big regression.

We've addressed the underlying issue that caused messages to be written to the database out of order, but you're probably still experiencing the lingering effects because your team members are still acting based on invalid or incomplete data.

### What you can do about it

Drives are loaded in the background using a browser feature called "SharedWorkers". This allows you to open new documents more quickly without having to load your personal and team drives every time. It also means that as long as you have any CryptPad tabs open your SharedWorker will remain active in the background. As a precaution for what I'm going to recommend next, you may want to tell your team members to close all their tabs to terminate any long-lived workers that may still be acting on the invalid fork of the drive.

We've implemented a small script that you can run from your browser's console. It will generate 100 arbitrary changes in the target drive and result in two checkpoints, ensuring that anyone who loads the document afterward ends up on the correct branch.


If you notice the issue in your own drive due to multiple devices in use at once:

1. open a CryptPad tab
2. wait for it to finish loading
3. open your browser's developer console
4. paste `Cryptpad.fixFork();` into the console and hit enter
5. wait approximately two minutes while it generates and sends arbitrary drivve modifications to the server

If the issue affects a team (as in your case), one and only one team member can call the same script with the affected team's id. There is another function (`Cryptpad.getTeamsId()`)  which will list team ids like so:

```js
/*
raccoons : 1519627469655575
CryptPad Squad : 4520771235786924
*/
```

Then run the "fixFork" command with the resulting number in the parentheses, like so: 

```js
// Fix the "raccoons" team
Cryptpad.fixFork(1519627469655575)
```


### In conclusion

Sorry for the trouble and thanks for your patience!
These measures were an effort to make a fix available as quickly as possible. We're actively discussing what we can do to make CryptPad more resilient against similar situations in the future.

**Comment by thomaskuntzz** [comment]
Wow, thanks for such a detailed answer!

Indeed things get clearer with this in-depth view of how the issue might occur... However if I understand well what you said, the issue should also occur on text pads for instance, since you said they share the same underlying structure as team drives. Why doesn't it happen?

I ran the code you said and have not experienced the bug since. However the `Cryptpad.fixFork(teamId)` call returned instantaneously (returned `undefined`) instead of running for about two minutes as you said it would. Is this normal? 🤔 ⌛

Thanks again for your help and all the time you put into writing this detailed answer! 👍

**Comment by ansuz** [comment]
You're correct that pads and team drives share the same type of data structure and are susceptible to the same underlying issue. In practice, however, it's far more likely to affect drives because the problem is exacerbated by long-lived sessions. Clients maintain an active session in the background as long as it has any tabs open, whereas a pad's session is terminated once its tab has been closed. That's true for every member of a team, so the problem could have occurred as long as at least one account has an active session with a conflicting version of the drive state. This situation would have become increasingly likely the more active members a team had.

**Comment by thomaskuntzz** [comment]
Cool, thanks for all those details! If you're confident that the underlying problem has been fixed then I think we can close this issue, as the problem hasn't occurred for me since I applied the fix you suggested 🙂

**Comment by ansuz** [comment]
Great!
